src/iceberg/datum.h

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include <compare>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+#include "iceberg/result.h"
+#include "iceberg/type.h"
+
+namespace iceberg {
+
+/// \brief Exception type for values that are below the minimum allowed value for a
+/// primitive type.
+///
+/// When casting a value to a narrow primitive type, if the value exceeds the maximum of
+/// dest type, it might be above the maximum allowed value for that type.
+struct BelowMin {
+  bool operator==(const BelowMin&) const = default;
+  std::strong_ordering operator<=>(const BelowMin&) const = default;
+};
+
+/// \brief Exception type for values that are above the maximum allowed value for a
+/// primitive type.
+///
+/// When casting a value to a narrow primitive type, if the value exceeds the maximum of
+/// dest type, it might be above the maximum allowed value for that type.
+struct AboveMax {
+  bool operator==(const AboveMax&) const = default;
+  std::strong_ordering operator<=>(const AboveMax&) const = default;
+};
+
+using PrimitiveLiteralValue =
+    std::variant<bool,                  // for boolean
+                 int32_t,               // for int, date
+                 int64_t,               // for long, timestamp, timestamp_tz, time
+                 float,                 // for float
+                 double,                // for double
+                 std::string,           // for string
```

Comment:
C++ string not requires utf8, so I wonder whether `std::vector<uint8_t>` with string is possible

Reply:
Small String Optimization (SSO) will make PrimitiveLiteralValue 40 bytes, may be use a unique_ptr to own the string?

Reply:
Emmm, a tiny wrapper like this should be defined if unique_ptr<string> should be added. I'm also ok for the case

```c++
struct StringWrapper {
public
  operator<=>();
  StringWrapper(StringWrapper&&);
  StringWrapper(const StringWrapper&);
private:
  std::unique_ptr<string>
};
```